tests/components/modbus/test_init.py

```diff
@@ -435,6 +436,25 @@ async def test_config_modbus(hass, caplog, mock_modbus_with_pymodbus):
     """Run configuration test for modbus."""
 
 
+async def test_failing_setup(hass):
+    """Test setup fails."""
+    config = {
+        DOMAIN: {
+            CONF_TYPE: TCP,
+            CONF_HOST: TEST_MODBUS_HOST,
+            CONF_PORT: TEST_PORT_TCP,
+            CONF_NAME: TEST_MODBUS_NAME,
+        },
+    }
+    assert await async_setup_component(hass, DOMAIN, config) is True
+    hub = hass.data[DOMAIN][TEST_MODBUS_NAME]
+    hub.async_setup = mock.AsyncMock(return_value=False)
+    setup_config = {
+        DOMAIN: [],
+    }
+    await async_modbus_setup(hass, setup_config)
```

Comment:
Here you are testing the `async_modbus_setup` function itself, not the flow that calls this function.
It seems to me that you should be test instead an integration reload with a failure to connect on reload.

```python
@pytest.mark.parametrize("do_config", [{}])
async def test_integration_reload_failed(
    hass: HomeAssistant, caplog, mock_modbus
) -> None:
    """Test setup fails."""
    caplog.set_level(logging.INFO)
    caplog.clear()

    yaml_path = get_fixture_path("configuration.yaml", "modbus")
    with mock.patch.object(
        hass_config, "YAML_CONFIG_FILE", yaml_path
    ), mock.patch.object(mock_modbus, "connect", side_effect=ModbusException("error")):
        await hass.services.async_call(DOMAIN, SERVICE_RELOAD, blocking=True)
        await hass.async_block_till_done()

    assert "Modbus reloading" in caplog.text
    assert "connect failed, retry in pymodbus" in caplog.text
```

Reply:
the problem with your example is that async_modbus_setup() is called with
```
config = 'modbus': [{'type': 'tcp', 'host': 'modbusHost', 'port': 5501, 'name': 'modbusTest', 'timeout': 3.0, 'delay': 0, 'retry_on_empty': False, 'retries': 3, 'close_comm_on_error': True}]}
```
Which in my mind is an error in Core, but I stopping arguing about that.

But due to the fact the config is not empty the code never enters
```
if DOMAIN in hass.data and config[DOMAIN] == []:
        hubs = hass.data[DOMAIN]
        for name in hubs:
            if not await hubs[name].async_setup():
                return False
```

I have been playing with many different scenarios to provoke an empty config, but without success.

The test I have added is solely to test what happens with async_setup() delivers False, and I think the test is good for that purpose.

Reply:
Did you try my test?
It does enter with empty config, and it does go through that path and completes 100% coverage.

It is the same as `test_integration_reload` (at the end of the test file) except that `mock_modbus.connect` triggers side_effect `ModbusException`

Reply:
Yes I copied your test into test_init.py, and ran it in VC, and the config was from the debug window. 

I will try your test again.